File: golang-leetcode/98.go
package main

import "fmt"

func main() {
	fmt.Println("vim-go")
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return isValid(root, nil, nil)
}

func isValid(root, min, max *TreeNode) bool {
	if root == nil {
		return true
	}
	if min != nil && root.Val <= min.Val {
		return false
	}
	if max != nil && root.Val >= max.Val {
		return false
	}
	return isValid(root.Left, min, root) &&
		isValid(root.Right, root, max)
}

File: python-leetcode/98.py
class TreeNode:

  def __init__(self, x):
    self.val = x
    self.left = None
    self.right = None


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:

  def isValidBST(self, root: TreeNode) -> bool:

    def dfs(root, minval, maxval):
      if root == None: return True
      if minval != None and root.val <= minval: return False
      if maxval != None and root.val >= maxval: return False
      return dfs(root.left, minval, root.val) and \
        dfs(root.right, root.val, maxval)

    return dfs(root, None, None)

node1 = TreeNode(0)
node2 = TreeNode(-1)
node1.right = node2
root = node1

s = Solution()
result = s.isValidBST(root)
print(result)
